ftl/python/builder.py

```diff
@@ -0,0 +1,150 @@
+# Copyright 2017 Google Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""This package defines the interface for orchestrating image builds."""
+
+import hashlib
+import os
+import subprocess
+import tempfile
+import logging
+import json
+import datetime
+
+from containerregistry.client.v2_2 import append
+from containerregistry.transform.v2_2 import metadata
+
+from ftl.common import builder
+
+_PYTHON_NAMESPACE = 'python-requirements-cache'
+_REQUIREMENTS_TXT = 'requirements.txt'
+_DEFAULT_TTL_WEEKS = 1
+
+
+class Python(builder.JustApp):
+    def __init__(self, ctx):
+        self._overrides = None
+        super(Python, self).__init__(ctx)
+
+    def __enter__(self):
+        """Override."""
+        return self
+
+    def CreatePackageBase(self, base_image, cache, use_cache=True):
+        """Override."""
+        # Figure out if we need to override entrypoint.
+        # Save the overrides for later to avoid adding an extra layer.
+        descriptor = None
+        for f in [_REQUIREMENTS_TXT]:
+            if self._ctx.Contains(f):
+                descriptor = f
+                descriptor_contents = self._ctx.GetFile(f)
+                break
+
+        if not descriptor:
+            logging.info('No package descriptor found. No packages installed.')
+        overrides = metadata.Overrides(
+            creation_time=str(datetime.date.today()) + "T00:00:00Z")
+        # TODO(aaron-prindle) currently if no requirements.txt, this fails here
+        checksum = hashlib.sha256(descriptor_contents).hexdigest()
+        if use_cache:
+            hit = cache.Get(base_image, _PYTHON_NAMESPACE, checksum)
+            if hit:
+                logging.info('Found cached dependency layer for %s' % checksum)
+                # TODO(aaron-prindle) check that cached dep layer is in TTL
```

Comment:
Drop the TODO? Maybe make this a separate function though, so we can re-use it in Node.js